### PR TITLE
APIMF-1961, APIMF-1978, APIMF-1979, APIMF-1980, APIMF-1981: Add Security Scheme warnings

### PR DESCRIPTION
**Pull request by arielmirra**
if the security scheme is of type "basic", it should not have extra settings fields such as tokenUrl or flow. Only valid fields are description and name.

### CODE REVIEW

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/Oas2SecuritySettingsParser.scala [hunk 1]
```diff
@@ -1,37 +1,53 @@
...
+              .map(entry => parseDynamicSettings(entry.value.as[YMap], defaultSettings))
+              .getOrElse(defaultSettings)
+        }
+        parsedSettings.annotations += VirtualObject()
+        parseAnnotations(parsedSettings)
       }
-      parsedSettings.annotations += VirtualObject()
-      parseAnnotations(parsedSettings)
-    }
+      .orElse({ handleNoSettings() })
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/Oas2SecuritySettingsParser.scala]: - [x] Remove parenthesis
- [x] Function name can be more descriptive like something with "validate"

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/Oas2SecuritySettingsParser.scala [hunk 2]
```diff
@@ -1,37 +1,53 @@
...
+  private def handleNoSettings(): Option[Settings] = {
+    val schemeType = scheme.`type`.value()
+
+    def settingsPresent: Boolean =
+      map.entries.exists(element => {
+        val field = element.key.as[String]
+        field != "description" && field != "name"
+      })
+
+    if (schemeType == "Basic Authentication" && settingsPresent)
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/Oas2SecuritySettingsParser.scala]:
```suggestion
   const BASIC_AUTHENTICATION_SCHEME = "Basic Authentication"
    if (schemeType == BASIC_AUTHENTICATION_SCHEME && settingsPresent)
```
Extract constant to a more significant place

File: amf-webapi/shared/src/main/scala/amf/validations/CustomShaclFunctions.scala [hunk 3]
```diff
@@ -231,13 +235,25 @@ object CustomShaclFunctions {
...
-          }
+          case p: Parameter =>
+            p.schema match {
+              case _: NodeShape => // ignore
+              case _            => violation(None)
+            }
+        }
+      }
+    }),
+    "mandatoryEmptyBasicSecurityScheme" -> ((element, violation) => {
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/validations/CustomShaclFunctions.scala]: - [x] Erase, no longer needed

File: amf-client/shared/src/test/resources/validations/oas2/security/missing-authorizationUrl-oauth2.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (tomsfernandez) [inline amf-client/shared/src/test/resources/validations/oas2/security/missing-authorizationUrl-oauth2.json]: - [x] Should be removed from commit if it isn't used.

**Review by tomsfernandez — CHANGES_REQUESTED** [review]

**Comment by tomsfernandez** [comment]
Talking with @AgustinBettati we thought that it would be worth analyzing if this could be implemented with a closedShape validation.

File: amf-client/shared/src/test/resources/validations/oas2/security/missing-authorizationUrl-oauth2.json (hunk 4 above)
Comment (arielmirra) [reply amf-client/shared/src/test/resources/validations/oas2/security/missing-authorizationUrl-oauth2.json]: it's not new, it's from a test that was on Oas30 and I moved it to Oas20 test suite to match its type.

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 5]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
-                                               entry: YPart,
-                                               isWarning: Boolean): Unit = {
-    val severity  = severities.getOrElse(shape, SeverityLevels.VIOLATION)
-    val isWarning = severity == SeverityLevels.WARNING
-    super.throwClosedShapeError(shape, node, message, entry, isWarning)
+  override def closedShape(node: String, ast: YMap, shape: String): Unit = {
+    syntax.nodes.get(shape) match {
+      case Some(properties) =>
+        ast.entries.foreach { entry =>
+          val key: String = entry.key.asScalar.map(_.text).getOrElse(entry.key.toString)
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: Should be refactored into a function getEntryKey() or something like that

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 6]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
+          if (ignore(shape, key)) {
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: shouldn't have empty clause. Should be merged with the clause below

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 7]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
+            // annotation or path in endpoint/webapi => ignore
+          } else if (!properties(key)) {
+            throwClosedShapeError(shape, node, s"Property '$key' not supported in a $vendor $shape node", entry)
+          }
+        }
+      case None =>
+        if (customSyntax.contains(shape)) {
+          val entries = ast.entries.map { entry =>
+            entry.key.asScalar.map(_.text).getOrElse(entry.key.toString)
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: Should be refactored into a function getEntryKey() or something like that

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 8]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
+    }
   }
 
+  private def checkRequired(node: String, ast: YMap, shape: String, entries: Seq[String]): Unit = {
+    if (customSyntax.contains(shape)) {
+      val specNode             = customSyntax(shape)
+      val (required, possible) = (specNode.requiredFields, specNode.possibleFields)
+
+      // if entries don't contain required fields
+      required.foreach { field =>
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: Refactor into function

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 9]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
+                                node,
+                                s"Property '${field.name}' is required in a $vendor $shape node",
+                                ast,
+                                isWarning)
+        }
+      }
+
+      // if invalid fields are present
+      entries.foreach(entry => {
+        if (!possible.contains(entry) && !required.map(_.name).contains(entry)) {
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: Possible entries should contain required entries. 
Avoid using `map` method for every entry.

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/OasSecuritySchemeParser.scala [hunk 10]
```diff
@@ -2,39 +2,82 @@ package amf.plugins.document.webapi.parser.spec.declaration
...
     ctx.closedShape(scheme.id, map, key)
+
+    val flow = map.key("flow").map(_.value.as[String]) match {
+      case Some("implicit")    => "implicit"
+      case Some("accessCode")  => "accessCode"
+      case Some("application") => "application"
+      case Some("password")    => "password"
+      case _                   => shape
+    }
+    if (!flow.isEmpty) ctx.closedShape(scheme.id, map, flow)
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/OasSecuritySchemeParser.scala]: only one closedShape call should run, not 2. If YAML map does not contain "flow" key then the **fallback** should be "oauth2"

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/OasSecuritySchemeParser.scala [hunk 11]
```diff
@@ -2,39 +2,82 @@ package amf.plugins.document.webapi.parser.spec.declaration
...
 
 case class OasSecuritySchemeParser(part: YPart, adopt: SecurityScheme => SecurityScheme)(
     implicit ctx: OasLikeWebApiContext)
     extends OasLikeSecuritySchemeParser(part, adopt) {
   override def closedShape(scheme: SecurityScheme, map: YMap, shape: String): Unit = {
+
     val key = map.key("type").map(_.value.as[String]) match {
-      case Some("basic") => "basic"
-      case _             => shape
+      case Some("basic")  => "basic"
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/OasSecuritySchemeParser.scala]: Key to call closedShape should be extracted into a function. Take into account the comment about **oauth2** being a fallback to also include the "flow" choosing

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 12]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
+          }
+          checkRequired(node, ast, shape, entries)
+        } else specificClosedShape(node, shape, ast)
+    }
   }
 
+  private def checkRequired(node: String, ast: YMap, shape: String, entries: Seq[String]): Unit = {
+    if (customSyntax.contains(shape)) {
+      val specNode             = customSyntax(shape)
+      val (required, possible) = (specNode.requiredFields, specNode.possibleFields)
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: Redundant, not necessary

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala [hunk 13]
```diff
@@ -28,17 +27,64 @@ class CustomClosedShapeContextDecorator(decorated: OasLikeWebApiContext,
...
+            throwClosedShapeError(shape, node, s"Property '$key' not supported in a $vendor $shape node", entry)
+          }
+        }
+      case None =>
+        if (customSyntax.contains(shape)) {
+          val entries = ast.entries.map { entry =>
+            entry.key.asScalar.map(_.text).getOrElse(entry.key.toString)
+          }
+          checkRequired(node, ast, shape, entries)
+        } else specificClosedShape(node, shape, ast)
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/CustomClosedShapeContextDecorator.scala]: specificClosedShape is redundant. Should use throwClosedShapeError

**Comment by tomsfernandez** [comment]
You should also limit the validations to OAS 2.0 as currently they are running for OAS 2.0 and OAS 3.0. To do this you should subclass the `OasSecuritySchemeParser` into Oas 2 and Oas 3 following convention and only apply the custom Syntax to OAS 2.0 only

**Comment by tomsfernandez** [comment]
Also to finish this feature you should run some sets just in case we are introducing a new validation that we didn't take into account.

**Review by tomsfernandez — CHANGES_REQUESTED** [review]

**Review by tomsfernandez — DISMISSED** [review]

**Review by AgustinBettati — APPROVED** [review]

**Pull request merged**